# HW6

In [1]:
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [112]:
# load L matrix
L = np.loadtxt("L.txt",dtype=np.float64)
print("L matrix\n",L)

L matrix
 [[-6.96069750e-09  6.94801823e-09  6.71236160e-12  5.96691455e-12]
 [ 1.22423101e-04 -1.96367157e-04  4.32134058e-05  3.07306501e-05]
 [ 4.36304686e-05  1.59415520e-02 -1.59851825e-02  0.00000000e+00]
 [ 4.22928296e-05  1.23619290e-02  0.00000000e+00 -1.24042218e-02]]


In [113]:
u,v = np.linalg.eig(L.T)

## (a)

In [114]:
# invariant distribution
pi = v[:,0]/np.sum(v[:,0])
print("Invariant distribution",pi)

Invariant distribution [9.99942954e-01 5.67509056e-05 1.53836961e-07 1.41077677e-07]


## (b)

In [115]:
u_right,v_right = np.linalg.eig(-L)
u_left,v_left = np.linalg.eig(-L.T)
print("Eigenvalues",u_right)
print("Columns as left eigenvectors\n",v_left)
print("Columns as right eigenvectors\n",v_right)

Eigenvalues [0.         0.00012202 0.01602898 0.01243478]
Columns as left eigenvectors
 [[ 9.99999998e-01  7.08944879e-01 -3.50923026e-03  4.55987104e-03]
 [ 5.67541431e-05 -7.05259068e-01  7.11845567e-01 -7.05059685e-01]
 [ 1.53845737e-07 -1.92122068e-03 -7.02301326e-01 -8.58156627e-03]
 [ 1.41085725e-07 -1.76459080e-03 -6.03501161e-03  7.09081381e-01]]
Columns as right eigenvectors
 [[ 5.00000000e-01  3.28131000e-05 -7.68693377e-10  9.07217631e-10]
 [ 5.00000000e-01 -5.75156324e-01  2.74744803e-03 -2.47165148e-03]
 [ 5.00000000e-01 -5.77998325e-01 -9.99952327e-01 -1.10978798e-02]
 [ 5.00000000e-01 -5.78889573e-01 -9.36992916e-03  9.99935362e-01]]


[-4.00375894e-09  7.01789003e-05  7.05256730e-05  7.06344205e-05]
[0.5 0.5 0.5 0.5]
[ 4.00375894e-09 -7.01789003e-05 -7.05256730e-05 -7.06344205e-05]


In [151]:
for k in range(1,4):
    print("Eigenvalue",u_right[k])
    F = np.zeros((4,4))
    for i in range(4):
        for j in range(4):
            F[i,j] = pi[i]*L[i,j]*(v_right[j,k] - v_right[i,k])
    print("F\n",F)
    ind = np.unravel_index(np.argmax(np.abs(F)),(4,4))
    print("Index with max. value of abs(F_ij)",ind[0]+1,ind[1]+1)
    
    for i in range(4):
        print("Eigencurrent at node",i+1," : ",-pi[i]*v_right[i,k])
    print("\n")

Eigenvalue 0.00012201708886640726
F
 [[-0.00000000e+00 -3.99619663e-09 -3.87973268e-12 -3.45418335e-12]
 [ 3.99619663e-09 -0.00000000e+00 -6.96972251e-12 -6.51075662e-12]
 [ 3.87973268e-12  6.96972251e-12 -0.00000000e+00 -0.00000000e+00]
 [ 3.45418335e-12  6.51075662e-12  0.00000000e+00 -0.00000000e+00]]
Index with max. value of abs(F_ij) 1 2
Eigencurrent at node 1  :  -3.2811228194463544e-05
Eigencurrent at node 2  :  3.264064229255034e-05
Eigencurrent at node 3  :  8.891750560527225e-08
Eigencurrent at node 4  :  8.166839630790653e-08


Eigenvalue 0.01602898317411897
F
 [[-0.00000000e+00  1.90882354e-11 -6.71165870e-12 -5.59063727e-14]
 [-1.90882354e-11 -0.00000000e+00 -2.45902084e-09 -2.11326116e-11]
 [ 6.71165870e-12  2.45902084e-09 -0.00000000e+00  0.00000000e+00]
 [ 5.59063727e-14  2.11326116e-11 -0.00000000e+00 -0.00000000e+00]]
Index with max. value of abs(F_ij) 2 3
Eigencurrent at node 1  :  7.686495259102252e-10
Eigencurrent at node 2  :  -1.5592016414032887e-07
Eigencurrent 

- For Eigenvalue = 0.0001222, it corresponds to transition process from 2 to 1, most eigencurrent is emitted at 2 and absorbed at 1
- For Eigenvalue = 0.01602, it corresponds to transition process from 3 to 2, most eigencurrent is emitted at 3 and absorbed at 2
- For Eigenvalue = 0.01243, it corresponds to transition process from 4 to 2, most eigencurrent is emitted at 4 and absorbed at 2

## (c)

In [134]:
ind_A = 1
ind_B = 0

A = np.array([[L[2,2],L[2,3]],[L[3,2],L[3,3]]])
b = np.array([-L[2,0],-L[3,0]])
q_23 = np.linalg.solve(A,b)
q = np.array([1,0,q_23[0],q_23[1]])
print("Forward Commitor values",q)
# calculate the reactive rate
mu = 0
for i in range(4):
    for j in range(4):
        F_ij = pi[i]*L[i,j]*(q[j] - q[i])
        if q[i] <= 0.5 and q[j] > 0.5:
            mu += F_ij
print("Reactive current",mu)
rho_A = np.sum(pi*(1-q))
k_AB = mu/rho_A
print("Escape rate",k_AB)

Forward Commitor values [1.         0.         0.00272943 0.00340955]
Reactive current 6.960261759438633e-09
Escape rate 0.00012201370138800253


The escape rate is close to the first eigenvalue above $\approx 0.00012201708886640726$ which corresponds to the transition process along minima 2 to 1.

## (d)

In [152]:
ind_A = 3
ind_B = 0

A = np.array([[L[1,1],L[1,2]],[L[2,1],L[2,2]]])
b = np.array([-L[1,0],-L[2,0]])
q_12 = np.linalg.solve(A,b)
q = np.array([1,q_12[1],q_12[1],0])
print("Forward Commitor values",q)

# calculate the reactive rate
mu = 0
for i in range(4):
    for j in range(4):
        F_ij = pi[i]*L[i,j]*(q[j] - q[i])
        #print(i,j,F_ij)
        # using an isocommitor cut
        if q[i] <= 0.5 and q[j] > 0.5:
            mu += F_ij
print("Reactive current",mu)
rho_A = np.sum(pi*(1-q))
k_AB = mu/rho_A
print("Escape rate",k_AB)

Forward Commitor values [1.         0.80004937 0.80004937 0.        ]
Reactive current 1.4012464538890519e-09
Escape rate 0.00012164424686739148


The escape rate is close to the first eigenvalue above $\approx 0.00012201708886640726$ which corresponds to the transition process along minima 2 to 1.

From the values of $F_{ij}$, the current is concentrated on the edges $4 \rightarrow 2, 2 \rightarrow 1$. 